src/communicator.h

```diff
@@ -38,6 +40,67 @@ class Communicator
  */
 virtual void initialize() = 0;
 
+/**
+ * Define the start point of a collective communication.
+ *
+ * Note: nested start/stop pairs are not supported and will lead to undefined behavior.
+ */
+virtual void start() = 0;
+
+/**
+ * Blocked until all communication since the *start* has been completed.
+ */
+virtual void stop() = 0;
+
+/**
+ * Send data to a remote rank.
+ *
+ * @param[in] buf           Data buffer to send to remote rank
+ * @param[in] count         Number of elements to send
+ * @param[in] element_size  Size of each element
+ * @param[in] dest          Destination rank
+ */
+virtual void send(const void *buf, int64_t count, int element_size, int dest) = 0;
+
+/**
+ * Receive data from a remote rank.
+ *
+ * @param[in] buf           Receive buffer to place received data into
+ * @param[in] count         Number of elements to receive
+ * @param[in] element_size  Size of each element
+ * @param[in] source        Source rank
+ */
+virtual void recv(void *buf, int64_t count, int element_size, int source) = 0;
+
+/**
+ * Close the endpoints, free up used communication resources, and stop the communication runtime.
+ */
+virtual void finalize() = 0;
+
+int mpi_rank;
+int mpi_size;
+int current_device;
+cudaStream_t comm_stream;
+
+};
+
+
+class MPILikeCommunicator : public Communicator
```

Comment:
Can you add some comments what do we need this communicator for? i.e. do we expect to have an MPI communicator deriving from this and sharing MPILike base class with the UCX communicator? Maybe also clarify it's a pure abstract class.